tq/transfer_queue.go

```diff
@@ -165,12 +196,266 @@ func (q *TransferQueue) Add(t Transferable) {
 		q.transferables[t.Oid()] = t
 		q.trMutex.Unlock()
 	} else {
-		tracerx.Printf("already transferring %q, skipping duplicate", t)
 		q.trMutex.Unlock()
+
+		tracerx.Printf("already transferring %q, skipping duplicate", t.Oid())
 		return
 	}
 
-	q.batcher.Add(t)
+	q.incoming <- t
+}
+
+// collectBatches collects batches in a loop, prioritizing failed items from the
+// previous before adding new items. The process works as follows:
+//
+//   1. Create a new batch, of size `q.batchSize`, and containing no items
+//   2. While the batch contains less items than `q.batchSize` AND the channel
+//      is open, read one item from the `q.incoming` channel.
+//      a. If the read was a channel close, go to step 4.
+//      b. If the read was a Transferable item, go to step 3.
+//   3. Append the item to the batch.
+//   4. Sort the batch by descending object size, make a batch API call, send
+//      the items to the `*adapterBase`.
+//   5. Process the worker results, incrementing and appending retries if
+//      possible.
+//   6. If the `q.incoming` channel is open, go to step 2.
+//   7. If the next batch is empty AND the `q.incoming` channel is closed,
+//      terminate immediately.
+//
+// collectBatches runs in its own goroutine.
+func (q *TransferQueue) collectBatches() {
+	var closing bool
+	batch := q.makeBatch()
+
+	for {
+		for !closing && !batch.IsFull(q.batchSize) {
+			t, ok := <-q.incoming
+			if !ok {
+				closing = true
+				break
+			}
+
+			batch = append(batch, t)
+		}
+
+		// Before enqueuing the next batch, sort it's items by size
+		// (largest first) so that workers can process the largest
+		// transfers first.
+		sort.Sort(sort.Reverse(batch))
```

Comment:
Sorry, I don't understand what benefit this sort is giving you. Since you're only sorting within the batch, and the batch is the unit that is blocked on, what does it matter what the ordering is inside the batch?

I could imagine that if you sorted all the transfers as a whole so that bigger files made it into the first batch that this would alter the blocking behaviour, but sorting inside the batch seems incidental. Please explain?